src/vm/functions/principals.rs

```diff
@@ -45,3 +48,76 @@ pub fn special_is_standard(
         (address_is_mainnet && context_is_mainnet) || (address_is_testnet && !context_is_mainnet),
     ))
 }
+
+pub fn special_parse_principal(
+    args: &[SymbolicExpression],
+    env: &mut Environment,
+    context: &LocalContext,
+) -> Result<Value> {
+    check_argument_count(2, args)?;
+    runtime_cost(ClarityCostFunction::Unimplemented, env, 0)?;
+
+    // Handle the block property name input arg.
+    let property_name = args[0]
+        .match_atom()
+        .ok_or(CheckErrors::ParsePrincipalExpectPropertyName)?;
+
+    let principal_property = PrincipalProperty::lookup_by_name(property_name).ok_or(
+        CheckErrors::NoSuchParsePrincipalProperty(property_name.to_string()),
+    )?;
+
+    let principal = eval(&args[1], env, context)?;
+
+    let (version, pub_key_hash) = match principal {
+        Value::Principal(PrincipalData::Standard(StandardPrincipalData(version, bytes))) => {
+            (version, bytes)
+        }
+        Value::Principal(PrincipalData::Contract(QualifiedContractIdentifier { issuer, name })) => {
+            (issuer.0, issuer.1)
+        }
+        _ => {
+            return Err(CheckErrors::TypeValueError(TypeSignature::PrincipalType, principal).into())
+        }
+    };
+
+    let result = match principal_property {
+        PrincipalProperty::Version => Value::UInt(version as u128),
+        PrincipalProperty::PubKeyHash => Value::Sequence(SequenceData::Buffer(BuffData {
+            data: pub_key_hash.into(),
+        })),
+    };
+    Ok(result)
+}
+
+pub fn native_assemble_principal(version: Value, pub_key_hash: Value) -> Result<Value> {
+    let verified_version = match version {
```

Comment:
Should you actually verify here that the `version` is valid?

Reply:
Pavitthra also said same thing. We can check this.